jetty-servlet/src/main/java/org/eclipse/jetty/servlet/DecoratingListener.java

```diff
@@ -0,0 +1,177 @@
+
+            // Check we have the right MethodTypes for the current Decorator signatures
+            MethodHandles.Lookup lookup = MethodHandles.lookup();
+            lookup.findVirtual(Decorator.class, "decorate", DECORATE_TYPE);
+            lookup.findVirtual(Decorator.class, "destroy", DESTROY_TYPE);
+        }
+        catch (Exception e)
+        {
+            throw new IllegalStateException(e);
+        }
+    }
+
+    private final ServletContextHandler _context;
+    private final String _attributeName;
+    private Decorator _decorator;
+
+    public DecoratingListener(ServletContextHandler context, String attributeName)
+    {
+        Objects.requireNonNull(context);
+        Objects.requireNonNull(attributeName);
+        _context = context;
+        _attributeName = attributeName;
+        Object decorator = _context.getAttribute(_attributeName);
+        if (decorator != null)
+            _context.getObjectFactory().addDecorator(asDecorator(decorator));
+    }
+
+    public String getAttributeName()
+    {
+        return _attributeName;
+    }
+
+    public ServletContext getServletContext()
+    {
+        return _context.getServletContext();
+    }
+
+    private Decorator asDecorator(Object object)
+    {
+        if (object == null)
+            return null;
+        if (object instanceof Decorator)
+            return (Decorator)object;
+
+        try
+        {
+            Class<?> clazz = object.getClass();
+
+            MethodHandles.Lookup lookup = MethodHandles.lookup();
+            MethodHandle decorate = lookup.findVirtual(clazz, "decorate", DECORATE_TYPE);
+            MethodHandle destroy = lookup.findVirtual(clazz, "destroy", DESTROY_TYPE);
+            return new DynamicDecorator(decorate, destroy, object);
+        }
+        catch (Exception e)
+        {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public void attributeAdded(ServletContextAttributeEvent event)
+    {
+        if (_attributeName.equals(event.getName()))
+        {
+            _decorator = asDecorator(event.getValue());
+            _context.getObjectFactory().addDecorator(_decorator);
+        }
+    }
+
+    @Override
+    public void attributeRemoved(ServletContextAttributeEvent event)
+    {
+        if (_attributeName.equals(event.getName()) && _decorator != null)
+        {
+            _context.getObjectFactory().removeDecorator(_decorator);
+            _decorator = null;
+        }
+    }
+
+    @Override
+    public void attributeReplaced(ServletContextAttributeEvent event)
+    {
+        attributeRemoved(event);
+        attributeAdded(event);
+    }
+
+    private static class DynamicDecorator implements Decorator
+    {
+        private final MethodHandle _decorate;
+        private final MethodHandle _destroy;
+        private final Object _object;
+
+        private DynamicDecorator(MethodHandle decorate, MethodHandle destroy, Object object)
```

Comment:
Could I convince you to match the java.invoke order of parameters? (where the object instance to be interacted with is always the first parameter)
I know this is super minor, but ... :-)

``` java
private DynamicDecorator(Object object, MethodHandle decorate, MethodHandle destroy)
```

Reply:
sure